File: src/core/assembly/spec_table.py
"""§4.4 双向细目表约束 schema（T-W4-027）.

架构 v2 §4.4「测量(预留)」行：「双向细目表（内容×认知×题量×难度）= 单元格计数
约束」。本模块落地该 schema 的纯数据模型与校验，供 T-W4-028 CP-SAT 求解器
编译为约束、T-W4-029 测量卷产出做合规校验。

双向细目表（Two-Way Specification Table）：
- 第一维：内容（知识点编码 content_code，可任意层级深度的点分树形 code）
- 第二维：认知层级（Bloom 六级，与 Objective.cognitive_level 同口径）
- 每个单元格 {target_count, difficulty_min, difficulty_max}：
  - target_count：该单元格目标题数
  - difficulty_min/max：题目难度区间（p_correct 口径，越大越易；与
    item_param.params.difficulty、AssemblyProfile.target_p_correct_range
    同口径，避免单位混估）

为什么 difficulty 用 p_correct 而非「难度系数（越大越难）」：
- 平台既有参数体系（ctt.py / bayesian_shrinkage.py / item_param.params.difficulty）
  一致采用「p_correct = 答对率，越大越易」；
- 双向细目表的难度区间要与候选池 p_correct_prior 直接比较，同单位才能编译为
  CP-SAT 约束（T-W4-028）；
- 命名「difficulty」沿用任务卡原文，但语义= p_correct 区间（注释明示）。

校验规则（任务卡验收标准 §2）：
1. 全部单元格 target_count 之和 > 0（空表无组卷意义）
2. 单个单元格 difficulty_min ≤ difficulty_max（区间合法）
3. 维度编码存在性：content_code 必须存在于所引用图谱的 kp_node.code 集合
   （validate_against_graph 显式调用，不在构造期强制——构造期不知图谱范围）

序列化（任务卡验收标准 §3）：
- JSON / YAML 序列化无损（Pydantic model_dump / model_validate 往返；
  YAML 经 JSON 中转，pyyaml 安全 load）

宪法 A5/A7：本模块不 import 任何学科包/学段包（学科零特判）；
content_code 是字符串编码，由调用方从学科包图谱加载后传入。
"""
from __future__ import annotations

from typing import Any, Iterable, Optional

import yaml
from pydantic import BaseModel, ConfigDict, Field, model_validator

# 认知层级六值（与 Objective.cognitive_level 同口径，避免单位混估）
CognitiveLevel = str  # 见下方 _COGNITIVE_LEVELS 校验
_COGNITIVE_LEVELS: frozenset[str] = frozenset(
    {"remember", "understand", "apply", "analyze", "evaluate", "create"}
)


class SpecCell(BaseModel):
    """双向细目表单元格：内容×认知 的目标题量与难度区间.

    Attributes:
        content_code: 知识点编码（任意层级点分树形 code，如
            "math.nal.decimal.compare"）；存在性由 SpecTable.validate_against_graph
            显式校验，构造期不强制。
        cognitive_level: 认知层级（Bloom 六级，与 Objective.cognitive_level 同集）。
        target_count: 该单元格目标题数（≥0；0 表示该单元格不要求题量）。
        difficulty_min: 题目难度下限（p_correct 口径，[0.0, 1.0]，越大越易）。
        difficulty_max: 题目难度上限（p_correct 口径，[0.0, 1.0]，越大越易）。
    """

    model_config = ConfigDict(extra="forbid")

    content_code: str = Field(min_length=1)
    cognitive_level: str
    target_count: int = Field(ge=0)
    difficulty_min: float = Field(ge=0.0, le=1.0)
    difficulty_max: float = Field(ge=0.0, le=1.0)

    @model_validator(mode="after")
    def _check_cognitive_level_domain(self) -> "SpecCell":
        if self.cognitive_level not in _COGNITIVE_LEVELS:
            raise ValueError(
                f"cognitive_level {self.cognitive_level!r} 越域；"
                f"合法域 {sorted(_COGNITIVE_LEVELS)}"
            )
        return self

    @model_validator(mode="after")
    def _check_difficulty_range(self) -> "SpecCell":
        # 任务卡验收标准 §2：单个单元格 difficulty_min ≤ difficulty_max
        if self.difficulty_min > self.difficulty_max:
            raise ValueError(
                f"单元格 [{self.content_code}/{self.cognitive_level}] "
                f"difficulty_min={self.difficulty_min} > "
                f"difficulty_max={self.difficulty_max}"
            )
        return self


class SpecTable(BaseModel):
    """双向细目表：单元格集合 + 引用元数据.

    一份 SpecTable = 一次测量卷的内容×认知×题量×难度目标分布。
    表本身只增不改（D1 风格，ORM 层物理强制；本 Pydantic 模型只负责 schema）。

    Attributes:
        spec_table_id: 表 id（与 ORM 主键一致；ULID 或语义 id）。
        spec_table_version: 表版本（同 id 改版需递增版本；D1 版本账）。
        gradeband: 学段（L/M/H，与 AssemblyProfile.gradeband 同集）。
        graph_release: 引用的知识图谱 release id（维度编码存在性校验依据）。
        cells: 单元格列表；(content_code, cognitive_level) 唯一。
        total_count: 派生字段，所有 cells 的 target_count 之和；构造期校验 >0。
    """

    model_config = ConfigDict(extra="forbid")

    spec_table_id: str = Field(min_length=1)
    spec_table_version: str = Field(min_length=1)
    gradeband: str
    graph_release: str = Field(min_length=1)
    cells: list[SpecCell] = Field(min_length=1)

    @model_validator(mode="after")
    def _check_gradeband_domain(self) -> "SpecTable":
        if self.gradeband not in {"L", "M", "H"}:
            raise ValueError(
                f"gradeband {self.gradeband!r} 越域；合法域 ['L', 'M', 'H']"
            )
        return self

    @model_validator(mode="after")
    def _check_total_count_positive(self) -> "SpecTable":
        # 任务卡验收标准 §2：全部单元格 target_count 之和 > 0
        total = self.total_count
        if total <= 0:
            raise ValueError(
                f"细目表所有单元格 target_count 之和 = {total}，必须 > 0"
            )
        return self

    @model_validator(mode="after")
    def _check_cell_uniqueness(self) -> "SpecTable":
        # 同一 (content_code, cognitive_level) 唯一：否则单元格语义重叠，组卷
        # 时无法判定题量归属。允许同一 content_code 在不同 cognitive_level 出现。
        seen: set[tuple[str, str]] = set()
        dupes: list[str] = []
        for c in self.cells:
            key = (c.content_code, c.cognitive_level)
            if key in seen:
                dupes.append(f"{c.content_code}/{c.cognitive_level}")
            seen.add(key)
        if dupes:
            raise ValueError(
                f"细目表单元格 (content_code, cognitive_level) 重复：{dupes}"
            )
        return self

    @property
    def total_count(self) -> int:
        """所有单元格 target_count 之和（派生量）."""
        return sum(c.target_count for c in self.cells)

    # ────────────────────────────────────────────────────────────────
    # 维度编码存在性校验（任务卡验收标准 §2：维度编码存在性校验，引用知识图谱）
    # ────────────────────────────────────────────────────────────────

    def validate_against_graph(
        self, valid_content_codes: Iterable[str]
    ) -> list[str]:
        """校验所有 cell.content_code 存在于给定图谱编码集合.

        为什么不在构造期强制：构造期 SpecTable 不持有图谱快照（图谱是独立
        版本化资产，graph_release 字段只是引用指针）；调用方在组卷前以
        ``graph_release`` 对应的 kp_node.code 集合调用本方法做存在性校验。
        本方法与 ORM 层 FK 不同：图谱节点本身是版本化的（graph_release），
        跨版本存在性需运行期校验，无法用静态 FK 表达。

        Args:
            valid_content_codes: 图谱 release 中的合法 content_code 集合
                （通常来自 KpNode.code 列查询）。

        Returns:
            不存在的 content_code 列表（空列表表示全部存在）。

        Raises:
            ValueError: 若存在未知编码（含未知编码列表详情）。
        """
        valid = set(valid_content_codes)
        unknown = sorted({c.content_code for c in self.cells if c.content_code not in valid})
        if unknown:
            raise ValueError(
                f"细目表引用了图谱 {self.graph_release!r} 中不存在的 content_code："
                f"{unknown}（共 {len(unknown)} 个）；请检查 graph_release 或 cells"
            )
        return unknown

    def cell_at(self, content_code: str, cognitive_level: str) -> Optional[SpecCell]:
        """按 (content_code, cognitive_level) 取单元格；不存在返回 None."""
        for c in self.cells:
            if c.content_code == content_code and c.cognitive_level == cognitive_level:
                return c
        return None

    # ────────────────────────────────────────────────────────────────
    # 序列化（任务卡验收标准 §3：JSON/YAML 无损）
    # ────────────────────────────────────────────────────────────────

    def to_json(self) -> str:
        """序列化为 JSON 字符串（确定性：sort_keys，ensure_ascii=False）.

        为什么 sort_keys=True：序列化无损 + 跨进程一致（指纹/比对友好），
        与 AssemblyProfile.digest() 同手法。
        """
        import json

        return json.dumps(
            self.model_dump(mode="json"),
            sort_keys=True,
            ensure_ascii=False,
        )

    @classmethod
    def from_json(cls, data: str) -> "SpecTable":
        """从 JSON 字符串反序列化（与 to_json 互逆）."""
        import json

        return cls.model_validate(json.loads(data))

    def to_yaml(self) -> str:
        """序列化为 YAML 字符串（经 JSON 中转，保留类型；sort_keys 确定性）.

        为什么经 JSON 中转：Pydantic model_dump 后含 Python 类型（如 int/float），
        pyyaml dump 时 int 1 可能被识为 '1' 字符串的不确定性链路；JSON dump
        先归一为 JSON 类型（int/float/str/list/dict/null），yaml.safe_load 回来
        仍是同型，无损往返。与既有项目 yaml 用法（registries 加载）一致。
        """
        import json

        json_obj = json.loads(self.to_json())
        return yaml.safe_dump(json_obj, sort_keys=True, allow_unicode=True)

    @classmethod
    def from_yaml(cls, data: str) -> "SpecTable":
        """从 YAML 字符串反序列化（与 to_yaml 互逆）."""
        obj = yaml.safe_load(data)
        return cls.model_validate(obj)

    def to_dict(self) -> dict[str, Any]:
        """序列化为 dict（给 ORM 层 JSONB 落库用；与 model_dump(mode='json') 一致）."""
        return self.model_dump(mode="json")


__all__ = [
    "CognitiveLevel",
    "SpecCell",
    "SpecTable",
]

File: core/assembly/spec_table.go
// spec_table.go 承载双向细目表约束 schema（Python 冻结基准
// src/core/assembly/spec_table.py 的 Go 移植；架构 v2 §4.4「测量(预留)」行：
// 「双向细目表（内容×认知×题量×难度）= 单元格计数约束」）。
//
// 双向细目表（Two-Way Specification Table）：
//   - 第一维：内容（知识点编码 content_code，可任意层级深度的点分树形 code）
//   - 第二维：认知层级（Bloom 六级，与 Objective.cognitive_level 同口径）
//   - 每个单元格 {target_count, difficulty_min, difficulty_max}：
//     target_count 该单元格目标题数；difficulty_min/max 题目难度区间
//     （p_correct 口径，越大越易；与 item_param.params.difficulty、
//     AssemblyProfile.target_p_correct_range 同口径，避免单位混估）
//
// 为什么 difficulty 用 p_correct 而非「难度系数（越大越难）」：平台既有参数
// 体系一致采用「p_correct = 答对率，越大越易」；双向细目表的难度区间要与
// 候选池 p_correct_prior 直接比较，同单位才能编译为约束。命名沿用任务卡原文，
// 语义 = p_correct 区间。
//
// 校验规则：1) 全部单元格 target_count 之和 > 0；2) 单个单元格
// difficulty_min ≤ difficulty_max；3) 维度编码存在性（ValidateAgainstGraph
// 显式调用，不在构造期强制——构造期不知图谱范围）。
package assembly

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

// 认知层级六值（与 Objective.cognitive_level 同口径，避免单位混估）.
var cognitiveLevels = []string{
	"remember", "understand", "apply", "analyze", "evaluate", "create",
}

// IsValidCognitiveLevel 认知层级域判定.
func IsValidCognitiveLevel(level string) bool {
	return containsStr(cognitiveLevels, level)
}

// SpecCell 双向细目表单元格：内容×认知 的目标题量与难度区间（Python SpecCell）.
type SpecCell struct {
	// ContentCode 知识点编码（任意层级点分树形 code，如
	// "math.nal.decimal.compare"）；存在性由 SpecTable.ValidateAgainstGraph
	// 显式校验，构造期不强制.
	ContentCode string
	// CognitiveLevel 认知层级（Bloom 六级，与 Objective.cognitive_level 同集）.
	CognitiveLevel string
	// TargetCount 该单元格目标题数（≥0；0 表示该单元格不要求题量）.
	TargetCount int
	// DifficultyMin/Max 题目难度上下限（p_correct 口径，[0.0,1.0]，越大越易）.
	DifficultyMin float64
	DifficultyMax float64
}

// validate 构造期校验（Python model_validator：认知层级域 + 难度区间 + 计数域）.
func (c *SpecCell) validate() error {
	if len(c.ContentCode) < 1 {
		return fmt.Errorf("assembly: spec cell content_code 不能为空（min_length=1）")
	}
	if !IsValidCognitiveLevel(c.CognitiveLevel) {
		return fmt.Errorf("assembly: cognitive_level %q 越域；合法域 %v", c.CognitiveLevel, cognitiveLevels)
	}
	if c.TargetCount < 0 {
		return fmt.Errorf("assembly: spec cell [%s/%s] target_count = %d 越域（ge=0）", c.ContentCode, c.CognitiveLevel, c.TargetCount)
	}
	if c.DifficultyMin < 0.0 || c.DifficultyMin > 1.0 || c.DifficultyMax < 0.0 || c.DifficultyMax > 1.0 {
		return fmt.Errorf("assembly: spec cell [%s/%s] 难度区间越域（[0.0, 1.0]）", c.ContentCode, c.CognitiveLevel)
	}
	// 单个单元格 difficulty_min ≤ difficulty_max
	if c.DifficultyMin > c.DifficultyMax {
		return fmt.Errorf("assembly: 单元格 [%s/%s] difficulty_min=%v > difficulty_max=%v",
			c.ContentCode, c.CognitiveLevel, c.DifficultyMin, c.DifficultyMax)
	}
	return nil
}

// SpecTable 双向细目表：单元格集合 + 引用元数据（Python SpecTable）。
// 一份 SpecTable = 一次测量卷的内容×认知×题量×难度目标分布。表本身只增不改
// （D1 风格，ORM 层物理强制；本模型只负责 schema）。
type SpecTable struct {
	// SpecTableID 表 id（ULID 或语义 id）.
	SpecTableID string
	// SpecTableVersion 表版本（同 id 改版需递增版本；D1 版本账）.
	SpecTableVersion string
	// Gradeband 学段（L/M/H，与 AssemblyProfile.gradeband 同集）.
	Gradeband string
	// GraphRelease 引用的知识图谱 release id（维度编码存在性校验依据）.
	GraphRelease string
	// Cells 单元格列表；(content_code, cognitive_level) 唯一.
	Cells []SpecCell
}

// NewSpecTable 构造并校验细目表（Python model_validator 全量：学段域、
// total_count>0、单元格唯一性、单元格域）.
func NewSpecTable(specTableID, specTableVersion, gradeband, graphRelease string, cells []SpecCell) (*SpecTable, error) {
	st := &SpecTable{
		SpecTableID:      specTableID,
		SpecTableVersion: specTableVersion,
		Gradeband:        gradeband,
		GraphRelease:     graphRelease,
		Cells:            append([]SpecCell(nil), cells...),
	}
	if len(specTableID) < 1 || len(specTableVersion) < 1 || len(graphRelease) < 1 {
		return nil, fmt.Errorf("assembly: spec_table_id/spec_table_version/graph_release 不能为空（min_length=1）")
	}
	if !containsStr([]string{GradebandL, GradebandM, GradebandH}, gradeband) {
		return nil, fmt.Errorf("assembly: gradeband %q 越域；合法域 ['L', 'M', 'H']", gradeband)
	}
	if len(st.Cells) < 1 {
		return nil, fmt.Errorf("assembly: 细目表 cells 不能为空（min_length=1）")
	}
	for i := range st.Cells {
		if err := (&st.Cells[i]).validate(); err != nil {
			return nil, err
		}
	}
	// 全部单元格 target_count 之和 > 0
	if st.TotalCount() <= 0 {
		return nil, fmt.Errorf("assembly: 细目表所有单元格 target_count 之和 = %d，必须 > 0", st.TotalCount())
	}
	// 同一 (content_code, cognitive_level) 唯一：否则单元格语义重叠，组卷时
	// 无法判定题量归属。允许同一 content_code 在不同 cognitive_level 出现。
	type cellKey struct{ content, cognitive string }
	seen := map[cellKey]struct{}{}
	dupes := []string{}
	for _, c := range st.Cells {
		k := cellKey{c.ContentCode, c.CognitiveLevel}
		if _, ok := seen[k]; ok {
			dupes = append(dupes, fmt.Sprintf("%s/%s", c.ContentCode, c.CognitiveLevel))
		}
		seen[k] = struct{}{}
	}
	if len(dupes) > 0 {
		return nil, fmt.Errorf("assembly: 细目表单元格 (content_code, cognitive_level) 重复：%v", dupes)
	}
	return st, nil
}

// TotalCount 所有单元格 target_count 之和（派生量；Python total_count）.
func (st *SpecTable) TotalCount() int {
	total := 0
	for i := range st.Cells {
		total += st.Cells[i].TargetCount
	}
	return total
}

// ValidateAgainstGraph 校验所有 cell.content_code 存在于给定图谱编码集合
// （Python validate_against_graph）。
//
// 为什么不在构造期强制：构造期 SpecTable 不持有图谱快照（图谱是独立版本化
// 资产，graph_release 字段只是引用指针）；调用方在组卷前以 graph_release 对应
// 的 kp_node.code 集合调用本方法做存在性校验。本方法与 ORM 层 FK 不同：图谱
// 节点本身是版本化的（graph_release），跨版本存在性需运行期校验。
//
// 返回不存在的 content_code 列表（全部存在时为空且 error=nil；存在未知编码时
// 返回排序后的未知编码列表并携带 error）。
func (st *SpecTable) ValidateAgainstGraph(validContentCodes []string) ([]string, error) {
	valid := map[string]struct{}{}
	for _, c := range validContentCodes {
		valid[c] = struct{}{}
	}
	unknownSet := map[string]struct{}{}
	for i := range st.Cells {
		if _, ok := valid[st.Cells[i].ContentCode]; !ok {
			unknownSet[st.Cells[i].ContentCode] = struct{}{}
		}
	}
	if len(unknownSet) == 0 {
		return []string{}, nil
	}
	unknown := make([]string, 0, len(unknownSet))
	for c := range unknownSet {
		unknown = append(unknown, c)
	}
	sort.Strings(unknown)
	return unknown, fmt.Errorf(
		"assembly: 细目表引用了图谱 %q 中不存在的 content_code：%v（共 %d 个）；请检查 graph_release 或 cells",
		st.GraphRelease, unknown, len(unknown))
}

// CellAt 按 (content_code, cognitive_level) 取单元格；不存在返回 nil
// （Python cell_at）.
func (st *SpecTable) CellAt(contentCode, cognitiveLevel string) *SpecCell {
	for i := range st.Cells {
		if st.Cells[i].ContentCode == contentCode && st.Cells[i].CognitiveLevel == cognitiveLevel {
			return &st.Cells[i]
		}
	}
	return nil
}

// dump 转 Python model_dump(mode="json") 同形值树（键名 snake_case 逐字一致）.
func (st *SpecTable) dump() map[string]any {
	cells := make([]any, 0, len(st.Cells))
	for i := range st.Cells {
		c := &st.Cells[i]
		cells = append(cells, map[string]any{
			"content_code":    c.ContentCode,
			"cognitive_level": c.CognitiveLevel,
			"target_count":    c.TargetCount,
			"difficulty_min":  c.DifficultyMin,
			"difficulty_max":  c.DifficultyMax,
		})
	}
	return map[string]any{
		"spec_table_id":      st.SpecTableID,
		"spec_table_version": st.SpecTableVersion,
		"gradeband":          st.Gradeband,
		"graph_release":      st.GraphRelease,
		"cells":              cells,
	}
}

// fromDump 从 Python model_dump 同形值树还原（JSON/YAML 反序列化共用前置）.
func specTableFromDump(obj map[string]any) (*SpecTable, error) {
	id, _ := obj["spec_table_id"].(string)
	version, _ := obj["spec_table_version"].(string)
	gradeband, _ := obj["gradeband"].(string)
	release, _ := obj["graph_release"].(string)
	rawCells, _ := obj["cells"].([]any)
	cells := make([]SpecCell, 0, len(rawCells))
	for _, rc := range rawCells {
		m, ok := rc.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("assembly: spec_table.cells 元素形状非法")
		}
		c := SpecCell{}
		c.ContentCode, _ = m["content_code"].(string)
		c.CognitiveLevel, _ = m["cognitive_level"].(string)
		n, err := asInt(m["target_count"])
		if err != nil {
			return nil, fmt.Errorf("assembly: spec cell target_count 非法: %v", err)
		}
		c.TargetCount = n
		dmin, err := asFloat(m["difficulty_min"])
		if err != nil {
			return nil, fmt.Errorf("assembly: spec cell difficulty_min 非法: %v", err)
		}
		dmax, err := asFloat(m["difficulty_max"])
		if err != nil {
			return nil, fmt.Errorf("assembly: spec cell difficulty_max 非法: %v", err)
		}
		c.DifficultyMin, c.DifficultyMax = dmin, dmax
		cells = append(cells, c)
	}
	return NewSpecTable(id, version, gradeband, release, cells)
}

// ToJSON 序列化为 JSON 字符串（Python to_json：确定性 sort_keys，
// ensure_ascii=False）。规范化与 Python json.dumps 逐字节对齐（canon.go），
// 跨实现可比对/互验。
func (st *SpecTable) ToJSON() string {
	return canonicalJSON(st.dump())
}

// FromJSON 从 JSON 字符串反序列化（与 ToJSON 互逆；Python from_json）.
func FromJSON(data string) (*SpecTable, error) {
	var obj map[string]any
	if err := json.Unmarshal([]byte(data), &obj); err != nil {
		return nil, fmt.Errorf("assembly: spec_table JSON 反序列化失败: %w", err)
	}
	return specTableFromDump(obj)
}

// ToYAML 序列化为 YAML 字符串（Python to_yaml：经 JSON 中转保留类型，
// sort_keys 确定性；Go 侧 yaml.Marshal 对 map[string]any 同样按键排序输出）.
func (st *SpecTable) ToYAML() (string, error) {
	out, err := yaml.Marshal(st.dump())
	if err != nil {
		return "", fmt.Errorf("assembly: spec_table YAML 序列化失败: %w", err)
	}
	return string(out), nil
}

// FromYAML 从 YAML 字符串反序列化（与 ToYAML 互逆；Python from_yaml：
// yaml.safe_load → model_validate）.
func FromYAML(data string) (*SpecTable, error) {
	var obj map[string]any
	dec := yaml.NewDecoder(strings.NewReader(data))
	dec.KnownFields(false) // YAML 别名/锚点展开后键域与 dump 一致；校验由 NewSpecTable 承担
	if err := dec.Decode(&obj); err != nil {
		return nil, fmt.Errorf("assembly: spec_table YAML 反序列化失败: %w", err)
	}
	return specTableFromDump(obj)
}

// ToDict 序列化为 dict（给 ORM 层 JSONB 落库用；与 model_dump(mode='json')
// 一致——Go 形即 dump 值树）.
func (st *SpecTable) ToDict() map[string]any {
	return st.dump()
}
